rep-2011.rst

```diff
@@ -0,0 +1,1421 @@
+
+The data structure includes:
+
+- a list of field names and types, but not default values
+- a recursive list of field names and types for referenced types
+- an optional user-defined interface version, or 0 if not provided
+
+.. TODO::
+
+    Should the message name, including package and namespace, be part of this?
+    Consider a situation where you have the same data structure but two different type names, should those two instances have the same hash?
+    They are unique when paired with their type name, so it should be ok, but it is a bit weird, perhaps, since we're not using this type hash to check for wire compatibility between differently named types.
+
+.. TODO::
+
+    Related TODO, should we just use the TypeDescription described in a later section?
+    It's essentially the same thing, but it does include the type name.
+    I (wjwwood) am leaning in this direction.
+
+The resulting data structure is hashed using a standard SHA-1 method, resulting in a standard 160-bit (20-byte) hash value which is also generally known as a "message digest".
+This hash is combined with a type version hash standard version, which we will call the "ROS IDL Hashing Standard" or "RIHS", the first version of which will be ``RIHS1``.
+They are combined using an ``_`` symbol, resulting in a complete type version hash like ``RIHS1_<160-bit SHA-1 of data structure>``.
+This allows the tooling to know if a hash mismatch is due to a change in this standard (what is being hashed) or due to a difference in the interface types themselves.
+
+For now, the list of field names and their types are the only contributing factors, but in the future that could change, depending on which "annotations" are supported in ``.idl`` files.
+The "IDL - Interface Definition and Language Mapping" design document\ [2]_ describes which features of the OMG IDL standard are supported by ROS 2.
+If that is extended in the future, then this data structure may need to be updated, and if so the "type version hash standard version" will also need to be incremented.
+
+.. TODO::
+
+    Re-audit the supported features from OMG IDL according to the referenced design document, including the @key annotation and how it may impact this for the reference implementation.
+
+The optional user-defined interface version makes it possible to change the version hash of a message that only changed in "field semantics" (i.e. without changing field names or types), and therefore makes it possible to write "transfer functions" to handle semantic-only conversions between versions.
+There is currently no standard way to specify the user-defined interface version in either ``.msg`` or ``.idl`` files.
+
+.. TODO::
+
+    Remove the idea of the user-defined interface version or define how it can be supplied by the user in one or more of the idl file kinds.
```

Comment:
Perhaps just allow users to version via a recommended strategy of having a "versionN: true" field that is replaced for new versions? First thought was have it have a default value, or define a const, but I think based on current description of hashing logic, those two particular things should not change the hash. Inclined to agree that we don't need a special mechanism in this REP for "manual hash break".